ampd/src/handlers/starknet_verify_msg.rs

```diff
@@ -0,0 +1,655 @@
+where
+    C: StarknetClient,
+{
+    verifier: TMAddress,
+    voting_verifier: TMAddress,
+    rpc_client: C,
+    latest_block_height: Receiver<u64>,
+}
+
+impl<C> Handler<C>
+where
+    C: StarknetClient + Send + Sync,
+{
+    pub fn new(
+        verifier: TMAddress,
+        voting_verifier: TMAddress,
+        rpc_client: C,
+        latest_block_height: Receiver<u64>,
+    ) -> Self {
+        Self {
+            verifier,
+            voting_verifier,
+            rpc_client,
+            latest_block_height,
+        }
+    }
+
+    fn vote_msg(&self, poll_id: PollId, votes: Vec<Vote>) -> MsgExecuteContract {
+        MsgExecuteContract {
+            sender: self.verifier.as_ref().clone(),
+            contract: self.voting_verifier.as_ref().clone(),
+            msg: serde_json::to_vec(&ExecuteMsg::Vote { poll_id, votes })
+                .expect("vote msg should serialize"),
+            funds: vec![],
+        }
+    }
+}
+
+#[async_trait]
+impl<V> EventHandler for Handler<V>
+where
+    V: StarknetClient + Send + Sync,
+{
+    type Err = Error;
+
+    async fn handle(&self, event: &events::Event) -> Result<Vec<Any>> {
+        let PollStartedEvent {
+            poll_id,
+            source_gateway_address,
+            messages,
+            participants,
+            expires_at,
+            contract_address,
+            ..
+        } = match event.try_into() as error_stack::Result<_, _> {
+            Err(report) if matches!(report.current_context(), EventTypeMismatch(_)) => {
+                return Ok(vec![]);
+            }
+            event => event.change_context(DeserializeEvent)?,
+        };
+
+        if self.voting_verifier != contract_address {
+            return Ok(vec![]);
+        }
+
+        if !participants.contains(&self.verifier) {
+            return Ok(vec![]);
+        }
+
+        let latest_block_height = *self.latest_block_height.borrow();
+        if latest_block_height >= expires_at {
+            info!(poll_id = poll_id.to_string(), "skipping expired poll");
+            return Ok(vec![]);
+        }
+
+        let unique_msgs = messages
+            .iter()
+            .unique_by(|msg| msg.message_id.to_string())
+            .collect::<Vec<_>>();
+
+        // key is the message_id of the tx holding the event
+        let mut events: HashMap<FieldElementAndEventIndex, ContractCallEvent> =
+            join_all(unique_msgs.iter().map(|msg| {
+                self.rpc_client
+                    .get_events_by_hash_contract_call(msg.message_id.clone())
+            }))
+            .await
+            .into_iter()
+            .flatten()
+            .collect();
+
+        let mut votes = vec![];
+        for msg in unique_msgs {
+            match events.remove(&msg.message_id) {
+                Some(gateway_event) => {
+                    votes.push(verify_msg(&gateway_event, msg, &source_gateway_address));
+                }
+
+                None => votes.push(Vote::NotFound),
```

Comment:
why don't you just move this logic inside of the above map over unique_msgs? And then just generate the corresponding vote immediately after getting the event. There's no need to the `events` collection then. This seems a bit uneccessary to iterate twice like this, and to use mutable collections.

Reply:
It makes sense, yes. I mechanically changed the previous one.
It this better? ([commit](https://github.com/axelarnetwork/axelar-amplifier/pull/731/commits/3aee888d2dbba763b8e8b327490402723eef2421))